evol/helpers/checkpointing.py

```diff
@@ -0,0 +1,59 @@
+import json
+import pickle
+from datetime import datetime
+from typing import List
+
+from os import listdir
+from os.path import isdir, exists, join
+
+from evol import Individual
+
+
+def checkpoint(individuals: List[Individual], directory: str, method: str='pickle') -> None:
+    """Checkpoint a list of individuals.
+
+    :param individuals: List of individuals to checkpoint.
+    :param directory: Location to store the checkpoint. A new file is created for every checkpoint.
+    :param method: One of "pickle" or "json". For json, the chromosomes need to be json-serializable.
+    """
+    if not isdir(directory):
+        raise FileNotFoundError('Cannot write to "{}": is not a directory.'.format(directory))
+    filename = join(directory,
+                    datetime.now().strftime("%Y%m%d-%H%M%S.%f") + ('.pkl' if method == 'pickle' else '.json'))
+    if exists(filename):
+        raise FileExistsError('Cannot write to "{}": file exists!'.format(filename))
+    if method == 'pickle':
+        with open(filename, 'wb') as pickle_file:
+            pickle.dump(individuals, pickle_file)
+    elif method == 'json':
+        with open(filename, 'w') as json_file:
+            json.dump([individual.__dict__ for individual in individuals], json_file)
```

Comment:
So I've added the milliseconds to do just that. But I guess it is still possible to get duplicates if you are doing _really_ nothing between checkpoints. I agree that we should prevent that. I guess there are two options:

- Add a `uuid`
- Check the file does not exist. If it does, complain (I mean, really, you shouldn't checkpoint twice in a millisecond)

Reply:
I'm fine with `uuid`, it should be obvious enough why we are using it.

Reply:
A problem with that approach is that you no longer guarantee you load the last checkpoint when there are multiple from the same millisecond; uuid's don't sort nicely.

Reply:
So - I again see multiple options:

1. We use uuids and issue a warning when loading a checkpoint when there are multiple from the same millisecond.
2. We skip checkpointing if the file already exist, and issue a warning.
3. We check whether the file exists, and if so increment a counter. This counter can be used to properly sort checkpoints. _But_, since this counter will need a fixed number of digits (else sorting does not work) we will need to think about the problem when someone exhausts the maximum number of checkpoints in a milliseconds.